File: fsharp/bob/Bob.fs
module Bob

open System

let isAllUppercase (input: string) =
    let letters = input |> Seq.filter Char.IsLetter
    if Seq.isEmpty letters
    then false
    else letters |> Seq.forall Char.IsUpper

let (|YelledQuestion|Yelled|Question|Empty|Talking|) (input: string) =
    let input = input.Trim()

    if String.IsNullOrWhiteSpace input
    then Empty
    else
        match isAllUppercase input, input |> Seq.last = '?' with
        | true, true -> YelledQuestion
        | true, false -> Yelled
        | false, true -> Question
        | _ -> Talking

let response (input: string): string =
    match input with
    | YelledQuestion -> "Calm down, I know what I'm doing!"
    | Yelled ->  "Whoa, chill out!"
    | Question -> "Sure."
    | Empty -> "Fine. Be that way!"
    | Talking -> "Whatever."

File: typescript/bob/bob.ts

export function hey(message: string): string {
    const isQuestion = message.trim().slice(-1) === "?";
    const isYelling= /[A-Z]/.test(message) && message.toUpperCase() === message;
    const isSilence = message.trim().length === 0;
    
    if (isQuestion && isYelling) return "Calm down, I know what I'm doing!"
    else if (isYelling) return "Whoa, chill out!"
    else if (isQuestion) return "Sure."
    else if (isSilence) return "Fine. Be that way!"
    return "Whatever.";
}
